### PR TITLE
[STATISTICS-31] Add survivalProbability function for cont. dist.

### PR DESCRIPTION
**Pull request by benwtrent**
This commit adds a new function for all continuous
distributions `survivalProbability`.

While a naive implementation would simply be
`1-cumulativeProbability`, that would result
in loss of precision.

For many of the current cont. dist. a higher
precision survival probability is calculated

For others, it is simply `1-cumulativeProbability`

Many tests were added to verify the following:
- the precision of cumulativeProbability
- the precision of survivalProbability
- That survivalProbabiliy is near 1-cumulative
- That survival and cumulative probabilities are
  complementary

Through this development, certain distributions
were found lacking precision for their
cumulativeProbabilities and were improved.

Those were:

- CauchyDistribution
- ExponentialDistribution
- LogNormalDistribution
- PretoDistribution
- WeibullDistribution

Addresses Jira: STATISTICS-31

### CODE REVIEW

**Comment by aherbert** [comment]
This is currently failing the CI build on checkstyle issues. You can perform the same checks that are run by the CI build using the default maven target:
```
mvn
```

This will run the following checks:
```
mvn clean verify apache-rat:check clirr:check checkstyle:check pmd:check spotbugs:check javadoc:javadoc
```

You can run each target independently until it passes and then force push your changes to this branch. Thanks.

**Comment by benwtrent** [comment]
Ah, I bet the coverage is for survival function extreme cases, will cover those predicates.

**Comment by aherbert** [comment]
There is a method in `ContinuousDistributionAbstractTest.testOutsideSupport()` where you should add the complement of the cumulative probability test for the expected values 0 and 1 at the edge of the support. That will then hit coverage.

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/BetaDistributionTest.java [hunk 1]
```diff
@@ -160,6 +229,41 @@ private void checkCumulative(double alpha, double beta, double[] x, double[] cum
...
+    private void checkCumulativeSurvivalComplement(double alpha, double beta) {
+        final BetaDistribution d = new BetaDistribution(alpha, beta);
+        for (int i = 0; i < CUMULATIVE_TEST_POINTS.length; i++) {
+            double x = CUMULATIVE_TEST_POINTS[i];
+            Assertions.assertEquals(1, d.cumulativeProbability(x) + d.survivalProbability(x), 1e-8);
+        }
+    }
+
+    private void checkCumulativePrecision(double alpha, double beta, double value, double expected) {
+        final double tolerance = 1e-15;
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/BetaDistributionTest.java:248]: This tolerance is much larger than the expected which are below 5e-17. The tolerance should be lower.

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/BetaDistributionTest.java [hunk 2]
```diff
@@ -160,6 +229,41 @@ private void checkCumulative(double alpha, double beta, double[] x, double[] cum
...
+        BetaDistribution d = new BetaDistribution(alpha, beta);
+        TestUtils.assertEquals(
+                "cumulative probability not precise at " + value + " for a=" + alpha + " & b=" + beta,
+                d.cumulativeProbability(value),
+                expected,
+                tolerance);
+    }
+
+    private void checkSurvivalPrecision(double alpha, double beta, double value, double expected) {
+        final double tolerance = 1e-15;
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/BetaDistributionTest.java:258]: This tolerance is much larger than the expected which are below 5e-17. The tolerance should be lower.

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/BetaDistributionTest.java [hunk 3]
```diff
@@ -149,6 +152,72 @@ void testCumulative() {
...
+        checkCumulativePrecision(4.0, 5.0, 0.00001, 6.999776002800025e-19);
+        checkCumulativePrecision(5.0, 4.0, 0.0001, 5.598600119996539e-19);
+        checkCumulativePrecision(6.0, 2.0, 0.001, 6.994000000000028e-18);
+        checkCumulativePrecision(2.0, 6.0, 1e-9, 2.0999999930000014e-17);
+    }
+
+    /**
+     * Precision tests for verifying that survivalFunction calculates accurately in cases
+     * where 1-sf(x) is inaccurately 1
+     * */
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/BetaDistributionTest.java:210]: Change `* */` to `*/`

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/BetaDistributionTest.java [hunk 4]
```diff
@@ -149,6 +152,72 @@ void testCumulative() {
...
+        checkCumulativePrecision(5.0, 5.0, 0.0001, 1.2595800539968654e-18);
+        checkCumulativePrecision(4.0, 5.0, 0.00001, 6.999776002800025e-19);
+        checkCumulativePrecision(5.0, 4.0, 0.0001, 5.598600119996539e-19);
+        checkCumulativePrecision(6.0, 2.0, 0.001, 6.994000000000028e-18);
+        checkCumulativePrecision(2.0, 6.0, 1e-9, 2.0999999930000014e-17);
+    }
+
+    /**
+     * Precision tests for verifying that survivalFunction calculates accurately in cases
+     * where 1-sf(x) is inaccurately 1
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/BetaDistributionTest.java:209]: Add a period at the end of the sentence

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java [hunk 5]
```diff
@@ -46,6 +46,20 @@
  * makeInverseCumulativeTestPoints() -- arguments used to test inverse cdf
  * makeInverseCumulativeTestValues() -- expected inverse cdf values
  * <p>
+ * <p>
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java:49]: Remove extra `<p>` tag

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java [hunk 6]
```diff
@@ -46,6 +46,20 @@
...
+ * If the continuous distribution provides higher precision implementations of cumulativeProbability
+ * and/or survivalProbability, the following methods should be implemented to provide testing.
+ * To use these tests, calculate the cumulativeProbability and survivalProbability such that their naive
+ * complement is exceptionally close to `1` and consequently could lose precision due to floating point
+ * arithmetic.
+ *
+ * NOTE: The default high-precision threshold is 1e-15.
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java:56]: Add a `<p>` tag. The default tolerance should be changed to 1e-22.

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java [hunk 7]
```diff
@@ -46,6 +46,20 @@
...
+ *
+ * makeCumulativePrecisionTestPoints() -- high precision test inputs
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java:58]: Put this block in `<pre>` tags otherwise it will be rendered as a single sentence.

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java [hunk 8]
```diff
@@ -98,6 +127,34 @@
...
     public abstract double[] makeCumulativeTestValues();
 
+    /** Creates the default cumulative probability precision test input values */
+    public double[] makeCumulativePrecisionTestPoints() {
+        return new double[0];
+    }
+
+    /**
+     * Creates the default cumulative probability precision test expected values
+     * Note: The default threshold is 1e-15, any expected values with much higher precision may
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java:137]: Update tolerance and add a period at the end of the sentence.

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java [hunk 9]
```diff
@@ -98,6 +127,34 @@
...
+
+    /** Creates the default survival probability precision test input values */
+    public double[] makeSurvivalPrecisionTestPoints() {
+        return new double[0];
+    }
+
+    /**
+     * Creates the default survival probability precision test expected values
+     * Note: The default threshold is 1e-22, any expected values with much higher precision may
+     *        not test the desired results without increasing precision threshold
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java:152]: No need for the extra space here before `not`

Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java:152]: No need for the extra space here before `not`

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java [hunk 10]
```diff
@@ -191,6 +252,54 @@ protected void verifyCumulativeProbabilities() {
...
+            TestUtils.assertEquals("survival + cumulative probability were not close to 1.0" +
+                    x, 1.0,
+                distribution.survivalProbability(x) + distribution.cumulativeProbability(x),
+                getTolerance());
+        }
+    }
+
+    /**
+     * Verifies that survival is simply not 1-cdf by testing calculations that would underflow that calculation and
+     * result in an inaccurate answer
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java:275]: Add a period to the end of the sentence.

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java [hunk 11]
```diff
@@ -496,6 +627,10 @@ protected double getTolerance() {
         return tolerance;
     }
 
+    protected double getHighPrecisionTolerance() {
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/ContinuousDistributionAbstractTest.java:630]: Add a simple javadoc for this

**Review by aherbert — CHANGES_REQUESTED** [review]
All the formulas look good. There are a few nits with formatting. There are some leftover references to the original high precision tolerance of 1e-15. The beta distribution is not really testing high precision due to this. I changed it to 1e-22 and the test still passed.

All I would change would be to remove the generation of a symmetric value x around the mean / median for the Normal and Cauchy distributions and implement the logic directly.
